[4.6] - Caused by: java.lang.ClassCastException: org.codehaus.groovy.runtime.GStringImpl cannot be cast to java.lang.String

## Description:
When upgrading from 4.5 to 4.6, I cannot run tests: `Caused by: java.lang.ClassCastException: org.codehaus.groovy.runtime.GStringImpl cannot be cast to java.lang.String`

## Error:
Bug PR: https://github.com/jaredsburrows/gradle-checker-framework-plugin/pull/6
```
13 tests completed, 6 failed
:test FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/travis/build/[secure]/gradle-checker-framework-plugin/build/reports/tests/test/index.html
* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':test'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:103)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:73)
	at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
	at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter.execute(ResolveBuildCacheKeyExecuter.java:66)
	at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
	at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
	at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
	at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:256)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:249)
	at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:238)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:663)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:597)
	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.api.GradleException: There were failing tests. See the report at: file:///home/travis/build/[secure]/gradle-checker-framework-plugin/build/reports/tests/test/index.html
	at org.gradle.api.tasks.testing.AbstractTestTask.handleTestFailures(AbstractTestTask.java:611)
	at org.gradle.api.tasks.testing.AbstractTestTask.executeTests(AbstractTestTask.java:483)
	at org.gradle.api.tasks.testing.Test.executeTests(Test.java:561)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:788)
	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:755)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:124)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
	at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:113)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:95)
	... 31 more
* Get more help at https://help.gradle.org
BUILD FAILED in 1m 46s
13 actionable tasks: 12 executed, 1 up-to-date
Publishing build scan...
https://gradle.com/s/elzds3h4ulfvs
The command "./gradlew clean build -s --scan" exited with 1.
cache.2
store build cache
0.01s
60.67schange detected (content changed, file is created, or file is deleted):
/home/travis/.gradle/caches/4.6/file-changes/last-build.bin
/home/travis/.gradle/caches/4.6/fileHashes/fileHashes.bin
/home/travis/.gradle/caches/4.6/fileHashes/fileHashes.lock
/home/travis/.gradle/caches/4.6/fileHashes/resourceHashesCache.bin
/home/travis/.gradle/caches/4.6/generated-gradle-jars/generated-gradle-jars.lock
/home/travis/.gradle/caches/4.6/generated-gradle-jars/gradle-api-4.6.jar
/home/travis/.gradle/caches/4.6/generated-gradle-jars/gradle-test-kit-4.6.jar
/home/travis/.gradle/caches/4.6/scripts/5oubvvq1xz4i5gjqlqqo9jfhp/cp_dsl/cp_dsldf5583fde4f7f1f2f3f5ea117e2cdff1/cache.properties
/home/travis/.gradle/caches/4.6/scripts/5oubvvq1xz4i5gjqlqqo9jfhp/cp_dsl/cp_dsldf5583fde4f7f1f2f3f5ea117e2cdff1/cp_dsldf5583fde4f7f1f2f3f5ea117e2cdff1.lock
/home/travis/.gradle/caches/4.6/scripts/5oubvvq1xz4i5gjqlqqo9jfhp/cp_dsl/cp_dsldf5583fde4f7f1f2f3f5ea117e2cdff1/metadata/metadata.bin
/home/travis/.gradle/caches/4.6/scripts/5oubvvq1xz4i5gjqlqqo9jfhp/dsl/dsldf5583fde4f7f1f2f3f5ea117e2cdff1/
...
changes detected, packing new archive
.
.
.
.
.
.
.
.
uploading archive
Done. Your build exited with 1.
```

## Repro:
```
git clone https://github.com/jaredsburrows/gradle-checker-framework-plugin
cd gradle-checker-framework-plugin
gradlew test
```

--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/AnnotationProcessorPathFactory.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/AnnotationProcessorPathFactory.java
@@ -31,11 +31,12 @@ import java.io.File;
 import java.util.Collections;
 import java.util.LinkedHashSet;
 import java.util.List;
+import java.util.Map;
 import java.util.Set;
 
 
 public class AnnotationProcessorPathFactory {
-    public static final String COMPILE_CLASSPATH_DEPRECATION_MESSAGE = "Putting annotation processors on the compile classpath";
+    public static final String COMPILE_CLASSPATH_DEPRECATION_MESSAGE = "The following annotation processors were detected on the compile classpath:";
     public static final String PROCESSOR_PATH_DEPRECATION_MESSAGE = "Specifying the processor path in the CompilerOptions compilerArgs property";
 
     private final FileCollectionFactory fileCollectionFactory;
@@ -142,8 +143,15 @@ public class AnnotationProcessorPathFactory {
                     if (hasExplicitProcessor) {
                         return compileClasspath.getFiles();
                     }
-                    if (!annotationProcessorDetector.detectProcessors(compileClasspath).isEmpty()) {
-                        DeprecationLogger.nagUserOfDeprecated(COMPILE_CLASSPATH_DEPRECATION_MESSAGE, "Please add them to the processor path instead. If these processors were unintentionally leaked on the compile classpath, use the -proc:none compiler option to ignore them.");
+                    Map<String, AnnotationProcessorDeclaration> processors = annotationProcessorDetector.detectProcessors(compileClasspath);
+                    if (!processors.isEmpty()) {
+                        DeprecationLogger.nagUserWith(
+                            COMPILE_CLASSPATH_DEPRECATION_MESSAGE +
+                            " '" + Joiner.on("' and '").join(processors.keySet()) + "'. " +
+                            "Detecting annotation processors on the compile classpath is deprecated and Gradle 5.0 will ignore them. " +
+                            "Please add them to the annotation processor path instead. " +
+                            "If you did not intend to use annotation processors, you can use the '-proc:none' compiler argument to ignore them."
+                        );
                         return compileClasspath.getFiles();
                     }
                     return Collections.emptySet();

--- a/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/AnnotationProcessorPathFactory.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/AnnotationProcessorPathFactory.java
@@ -30,6 +30,7 @@ import org.gradle.util.DeprecationLogger;
 import java.io.File;
 import java.util.Collections;
 import java.util.LinkedHashSet;
+import java.util.List;
 import java.util.Set;
 
 
@@ -55,7 +56,7 @@ public class AnnotationProcessorPathFactory {
      * @return An empty collection when annotation processing should not be performed, non-empty when it should.
      */
     public FileCollection getEffectiveAnnotationProcessorClasspath(final CompileOptions compileOptions, final FileCollection compileClasspath) {
-        if (compileOptions.getCompilerArgs().contains("-proc:none")) {
+        if (compileOptions.getAllCompilerArgs().contains("-proc:none")) {
             return fileCollectionFactory.empty("annotation processor path");
         }
         final FileCollection annotationProcessorPath = compileOptions.getAnnotationProcessorPath();
@@ -74,14 +75,15 @@ public class AnnotationProcessorPathFactory {
 
     private FileCollection getProcessorPathFromCompilerArguments(final CompileOptions compileOptions) {
         final FileCollection annotationProcessorPath = compileOptions.getAnnotationProcessorPath();
-        int pos = compileOptions.getCompilerArgs().indexOf("-processorpath");
+        List<String> compilerArgs = compileOptions.getAllCompilerArgs();
+        int pos = compilerArgs.indexOf("-processorpath");
         if (pos < 0) {
             return null;
         }
-        if (pos == compileOptions.getCompilerArgs().size() - 1) {
-            throw new InvalidUserDataException("No path provided for compiler argument -processorpath in requested compiler args: " + Joiner.on(" ").join(compileOptions.getCompilerArgs()));
+        if (pos == compilerArgs.size() - 1) {
+            throw new InvalidUserDataException("No path provided for compiler argument -processorpath in requested compiler args: " + Joiner.on(" ").join(compilerArgs));
         }
-        final String processorpath = compileOptions.getCompilerArgs().get(pos + 1);
+        final String processorpath = compilerArgs.get(pos + 1);
         if (annotationProcessorPath == null) {
             return fileCollectionFactory.fixed("annotation processor path", extractProcessorPath(processorpath));
         }
@@ -156,10 +158,11 @@ public class AnnotationProcessorPathFactory {
 
     private static boolean checkExplicitProcessorOption(CompileOptions compileOptions) {
         boolean hasExplicitProcessor = false;
-        int pos = compileOptions.getCompilerArgs().indexOf("-processor");
+        List<String> compilerArgs = compileOptions.getAllCompilerArgs();
+        int pos = compilerArgs.indexOf("-processor");
         if (pos >= 0) {
-            if (pos == compileOptions.getCompilerArgs().size() - 1) {
-                throw new InvalidUserDataException("No processor specified for compiler argument -processor in requested compiler args: " + Joiner.on(" ").join(compileOptions.getCompilerArgs()));
+            if (pos == compilerArgs.size() - 1) {
+                throw new InvalidUserDataException("No processor specified for compiler argument -processor in requested compiler args: " + Joiner.on(" ").join(compilerArgs));
             }
             hasExplicitProcessor = true;
         }

--- a/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java
@@ -37,6 +37,7 @@ import org.gradle.api.tasks.OutputDirectory;
 import org.gradle.api.tasks.PathSensitive;
 import org.gradle.api.tasks.PathSensitivity;
 import org.gradle.process.CommandLineArgumentProvider;
+import org.gradle.util.CollectionUtils;
 import org.gradle.util.DeprecationLogger;
 
 import javax.annotation.Nullable;
@@ -350,9 +351,7 @@ public class CompileOptions extends AbstractOptions {
     @Internal
     public List<String> getAllCompilerArgs() {
         ImmutableList.Builder<String> builder = ImmutableList.builder();
-        for (CharSequence arg : getCompilerArgs()) {
-            builder.add(arg.toString());
-        }
+        builder.addAll(CollectionUtils.stringize(getCompilerArgs()));
         for (CommandLineArgumentProvider compilerArgumentProvider : getCompilerArgumentProviders()) {
             builder.addAll(compilerArgumentProvider.asArguments());
         }

--- a/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java
+++ b/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java
@@ -350,7 +350,9 @@ public class CompileOptions extends AbstractOptions {
     @Internal
     public List<String> getAllCompilerArgs() {
         ImmutableList.Builder<String> builder = ImmutableList.builder();
-        builder.addAll(getCompilerArgs());
+        for (CharSequence arg : getCompilerArgs()) {
+            builder.add(arg.toString());
+        }
         for (CommandLineArgumentProvider compilerArgumentProvider : getCompilerArgumentProviders()) {
             builder.addAll(compilerArgumentProvider.asArguments());
         }

--- a/subprojects/language-java/src/test/groovy/org/gradle/api/tasks/compile/CompileOptionsTest.groovy
+++ b/subprojects/language-java/src/test/groovy/org/gradle/api/tasks/compile/CompileOptionsTest.groovy
@@ -213,4 +213,12 @@ class CompileOptionsTest extends Specification {
         then:
         compileOptions.bootClasspath == "lib2.jar"
     }
+
+    def "converts GStrings to Strings when getting all compiler arguments"() {
+        given:
+        compileOptions.compilerArgs << "Foo${23}"
+
+        expect:
+        compileOptions.allCompilerArgs.contains('Foo23')
+    }
 }

